Fix of heralds display (issue 207)

burlemarxiste opened the pull request:
This is a separate branch for this fix.

https://github.com/Quandela/Perceval/issues/207

ericbrts commented:
Hello @burlemarxiste ,
I've created an integration branch for UnitaryHACK 2024 contributions. I've just merged your Barrier PR. It seems it created conflicts. Could you please solve them before I review this PR?

I'm assigning the bounty to you right now.

burlemarxiste commented:
Sure! I will look into this tomorrow!

ericbrts commented:
Thanks, testing and reviewing it right now!

ericbrts commented on perceval/rendering/circuit/renderer.py:
```diff
@@ -43,12 +43,6 @@ class PortPos:
     def __init__(self, x, y, fixed=True):
```
I think you can remove the parameter `fixed=True`.

Also, `class PortPos` was meant to follow a herald "Port" moving along it's mode, following permutations. It seems `collect_herald_info` does it for both input and output, and replaces the need for any kind of "changing x and y on the fly". Is that correct?

burlemarxiste replied:
Sure, it's not needed, I can remove it.

ericbrts commented on perceval/rendering/circuit/renderer.py:
```diff
@@ -484,114 +575,217 @@ def _add_shape(self, lines, circuit, content, w, shape_fn=None, pos=None):
-        self._update_mode_style(lines, circuit, w)
-        for i in range(lines[0], lines[-1] + 1):
-            self._chart[i] += w
+        if w:
```
Was this added in order to sometimes use `Barrier` with 0 width?
Anyway, nice add.

burlemarxiste replied:
Yes, this was added for my initial implementation that had both "Identity" (acts like a barrier but invisible on the circuit), and "Barrier".

I don't think that's needed anymore, but if you think there might be a use case where some circuit elements have width=0, I can make sure it's used everywhere needed.

ericbrts commented on perceval/rendering/circuit/renderer.py:
```diff
@@ -484,114 +575,217 @@ def _add_shape(self, lines, circuit, content, w, shape_fn=None, pos=None):
+
+    def append_circuit(self, lines, circuit, content, pos=None):
+        w = self._skin.get_width(circuit)
+        self._add_shape(lines, circuit, content, w, pos=pos)
```
Maybe the `if w:` (from Renderer class) is missing here

ericbrts commented on perceval/rendering/circuit/renderer.py:
```diff
@@ -484,114 +575,217 @@ def _add_shape(self, lines, circuit, content, w, shape_fn=None, pos=None):
+
+    def append_subcircuit(self, lines, circuit, content):
+        w = self._skin.style_subcircuit['width']
+        self._add_shape(lines, circuit, content, w, self._skin.subcircuit_shape)
```
Maybe the `if w:` (from Renderer class) is missing here

ericbrts commented on tests/test_visualization.py:
```diff
@@ -322,8 +322,36 @@ def test_svg_processor_with_heralds_phys(tmp_path, save_figs):
+    _save_or_check(pc, tmp_path, sys._getframe().f_code.co_name, save_figs, recursive=True)
+
+
 def test_svg_processor_with_heralds_and_barriers_phys(tmp_path, save_figs):
```
This test fails in this branch now that heralds have moved (I think you predicted it in the `Barrier` branch :) )

Also, even when fixing the SVG reference, there seems to be a glitch where the zones left and right of the gray barrier are white (don't take the background color into account)
![image](https://github.com/Quandela/Perceval/assets/49883965/c2e02b10-51b6-448c-bef2-6dd4d8dc38ee)

Could you please fix this?

burlemarxiste replied:
That's strange, the reason why I added `set_background_color` to canvas was precisely to this: made an element aware of the background on which it is drawn. I'll look into it!

ericbrts reviewed (commented):
Thanks for this PR, you made the code better, I'd say!

I'm currently trying to make the pytests run on Github from this PR, hence the target branch change, but I'm kinda failing for now. Anyway, you can run `pytest tests/test_visualization.py` from your computer.

burlemarxiste commented:
I'm a bit lost with the branches at this moment. How should I proceed to continue making changes related to this issue, but with the barrier implementation included to it?

ericbrts commented:
You fix_heralds branch seems to be the correct source branch to work on. I'm messing with destination branches, because I want to enable the automated test workflow (which I failed to do). In the end, your synchronized with main + your Barrier branch, so working on fix_heralds is the way to go.

I'll merge it in an intermediate branch (right now, named "release/unitary-hack-2024") so that I can control to which scope we add these unitary hack contributions.

burlemarxiste commented:
Is it okay if I merge the latest version of my barrier branch into this? It would be easier to regenerate the figure for the test that fails. I might also add a few more tests (circuits with `recursive` True or False).

ericbrts commented:
The Barrier code I merged on Friday is already in your fix_heralds branch (or do you have newer code for Barriers?). I made you split your work into 3 different PR for easier reviews. Now we're reconstructing your work brick by brick. When you merged fix_heralds from the unitary-hack-2024 branch, you retrieved your work about Barriers:
![image](https://github.com/Quandela/Perceval/assets/49883965/93c1b80e-fba2-4f7c-8517-8bfa5989f7b6)

burlemarxiste commented:
Thank you for confirming! I have made the changes you suggested to check for null object `width`, updated the image, and fixed the problem with the missing background color (it was a line that disappeared when I merged with the barrier branch).

burlemarxiste commented:
I have found an issue with my implementation: the call to `flatten()` in `collect_herald_info` should be done with the same depth of recursion than the diagram is going to be drawn. Can I add an extra `recursion_level` argument to `flatten()` to fix this?

Can you confirm that this is how the diagram should look when `recursive=False`?

<img width="440" alt="Screenshot 2024-06-10 at 17 13 01" src="https://github.com/Quandela/Perceval/assets/171129116/1bd5a185-ee5a-4490-a5a7-ee1ee664e37a">

ericbrts commented:
Yes you can add a `recursion_level` parameter with a default value meaning max recursion level (like 0 ? -1 ?)

The picture for the non-recursive display is totally fine.

burlemarxiste commented:
Thanks for confirming that the image is correct.

I just understood something that answered my initial confusion about the `recursive` argument: **there are only 2 possible recursion levels.**  The reason is that when a Processor is added to a Processor, they are composed (no nesting between them); and iterating on a circuit always goes to the deepest level.

So there are only two possible levels, and there is no needed for an extra argument. I just need to collect the herald info at the component level when `recursive=False`, and circuit element level when `recursive=True`.

This latest commit should fix it.

ericbrts commented:
Thanks @burlemarxiste , I'm reviewing / testing again now. Do you wish to add other things to your PR, or can I merge once I approve it?

ericbrts commented:
Yes please, file as many issues as you wish!
